src/core/server/saved_objects/service/lib/update_objects_spaces.ts

```diff
@@ -0,0 +1,310 @@
+ * @public
+ */
+export interface SavedObjectsUpdateObjectsSpacesObject {
+  /** The type of the object to update */
+  id: string;
+  /** The ID of the object to update */
+  type: string;
+  /**
+   * The space(s) that the object to update currently exists in. This is only intended to be used by SOC wrappers.
+   *
+   * @internal
+   */
+  spaces?: string[];
+  /**
+   * The version of the object to update; this is used for optimistic concurrency control. This is only intended to be used by SOC wrappers.
+   *
+   * @internal
+   */
+  version?: string;
+}
+
+/**
+ * Options for the update operation.
+ *
+ * @public
+ */
+export interface SavedObjectsUpdateObjectsSpacesOptions extends SavedObjectsBaseOptions {
+  /** The Elasticsearch Refresh setting for this operation */
+  refresh?: MutatingOperationRefreshSetting;
+}
+
+/**
+ * The response when objects' spaces are updated.
+ *
+ * @public
+ */
+export interface SavedObjectsUpdateObjectsSpacesResponse {
+  objects: SavedObjectsUpdateObjectsSpacesResponseObject[];
+}
+
+/**
+ * Details about a specific object's update result.
+ *
+ * @public
+ */
+export interface SavedObjectsUpdateObjectsSpacesResponseObject {
+  /** The type of the referenced object */
+  type: string;
+  /** The ID of the referenced object */
+  id: string;
+  /** The space(s) that the referenced object exists in */
+  spaces: string[];
+  /** Included if there was an error updating this object's spaces */
+  error?: SavedObjectError;
+}
+
+// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
+type Left = { tag: 'Left'; error: SavedObjectsUpdateObjectsSpacesResponseObject };
+// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
+type Right = { tag: 'Right'; value: Record<string, any> };
+type Either = Left | Right;
+const isLeft = (either: Either): either is Left => either.tag === 'Left';
+const isRight = (either: Either): either is Right => either.tag === 'Right';
+
+export interface UpdateObjectsSpacesParams {
+  registry: ISavedObjectTypeRegistry;
+  allowedTypes: string[];
+  client: RepositoryEsClient;
+  serializer: SavedObjectsSerializer;
+  getIndexForType: (type: string) => string;
+  objects: SavedObjectsUpdateObjectsSpacesObject[];
+  spacesToAdd: string[];
+  spacesToRemove: string[];
+  options?: SavedObjectsUpdateObjectsSpacesOptions;
+}
+
+/**
+ * Gets all references and transitive references of the given objects. Ignores any object and/or reference that is not a multi-namespace
+ * type.
+ */
+export async function updateObjectsSpaces({
+  registry,
+  allowedTypes,
+  client,
+  serializer,
+  getIndexForType,
+  objects,
+  spacesToAdd,
+  spacesToRemove,
+  options = {},
+}: UpdateObjectsSpacesParams): Promise<SavedObjectsUpdateObjectsSpacesResponse> {
+  if (!spacesToAdd.length && !spacesToRemove.length) {
+    throw SavedObjectsErrorHelpers.createBadRequestError(
+      'spacesToAdd and/or spacesToRemove must be a non-empty array of strings'
+    );
+  }
```

Comment:
We _could_ do that, but then if you did this:
```ts
client.updateObjectsSpaces([{ type: 'invalid-type', id: '1' }], [], []);
```
It would return an empty response, while if you did this:
```ts
client.updateObjectsSpaces([{ type: 'invalid-type', id: '1' }], ['some-space'], []);
```
it would return a response with an object-not-found error for that result.

It seems the most consistent approach is to throw a bad request error here, IMO.